### PR TITLE
Cleaning up mesh and network watchers

### PR DESCRIPTION
**Pull request by nmittler**
This current file watcher code is duplicated between istiod and pilot. It also has hard-coded logic for updating each component that needs to know. This is a bit hacky and error prone, and will likely become a maintenance burden when more components need to know about updates to these resources.

This change makes the watchers for mesh config and networks a first-class citizen. Components that need to know about updates now register themselves and handle the events accordingly.

### CODE REVIEW

**Review by liamawhite — APPROVED** [review]
Not much to approve on the UX wg side of things!

**Review by myidpt — APPROVED** [review]

**Review by costinm — COMMENTED** [review]
Istiod (as separate component) is getting removed in separate PR, the env wg decided to use pilot for now.

Rest looks good in general, will take a closer look.

File: pilot/pkg/bootstrap/mesh.go [hunk 1]
```diff
@@ -16,174 +16,88 @@ package bootstrap
...
-	if err := s.initMeshConfiguration(args); err != nil {
-		return fmt.Errorf("mesh config: %v", err)
-	}
-	if err := s.initMeshNetworks(args); err != nil {
-		return fmt.Errorf("mesh networks: %v", err)
-	}
-	return nil
-}
+// initMesh creates the mesh in the pilotConfig from the input arguments.
+func newMeshWatcher(args *PilotArgs, kubeClient kubernetes.Interface,
```
Comment (costinm) [inline pilot/pkg/bootstrap/mesh.go]: I'm not sure what is changing here - but the current default is to use the file. 

It is critical to keep it this way since we still want to support working without k8s, and this is the 
most stable/tested code path.

Comment (costinm) [reply pilot/pkg/bootstrap/mesh.go]: (just looking for confirmation this is the intent of the change - hard to follow )

File: pilot/pkg/bootstrap/server.go [hunk 2]
```diff
@@ -147,22 +143,29 @@ func NewServer(args PilotArgs) (*Server, error) {
...
 
-	s := &Server{
-		fileWatcher: filewatcher.NewWatcher(),
-	}
-
-	prometheus.EnableHandlingTimeHistogram()
-
-	// Apply the arguments to the configuration.
-	if err := s.initKubeClient(&args); err != nil {
+	kubeClient, err := newKubeClient(&args)
```
Comment (costinm) [inline pilot/pkg/bootstrap/server.go]: Not sure if that was the code before - but we need to stop creating kube clients in every module, just reuse the existing one.

File: pilot/pkg/config/clusterregistry/multicluster.go [hunk 3]
```diff
@@ -128,19 +130,6 @@ func (m *Multicluster) DeleteMemberCluster(clusterID string) error {
 	return nil
 }
 
-// Hot reload mesh networks for remote clusters
-func (m *Multicluster) ReloadNetworkLookup(meshNetworks *meshconfig.MeshNetworks) {
```
Comment (costinm) [inline pilot/pkg/config/clusterregistry/multicluster.go]: Where did this go ?

File: pilot/pkg/config/coredatamodel/discovery.go [hunk 4]
```diff
@@ -79,17 +79,18 @@ func (d *MCPDiscovery) HandleCacheEvents(config model.Config, event model.Event)
 	defer d.cacheMutex.Unlock()
 	services := convertServices(config)
 	d.mergeCachedServices(services)
+	mixerEnabled := d.mixerEnabled()
 	switch event {
 	case model.EventAdd, model.EventUpdate:
-		newSvcInstancesByIP, newSvcInstancesByHost := d.convertInstances(config, services)
+		newSvcInstancesByIP, newSvcInstancesByHost := d.convertInstances(config, services, mixerEnabled)
```
Comment (costinm) [inline pilot/pkg/config/coredatamodel/discovery.go]: why do we need this ? Since mixer will go away soon I would avoid polluting the interfaces.

File: pilot/pkg/networking/core/v1alpha3/networkfilter.go [hunk 5]
```diff
@@ -79,8 +80,8 @@ func setAccessLog(env *model.Environment, node *model.Proxy, config *tcp_proxy.T
 
 // setAccessLogAndBuildTCPFilter sets the AccessLog configuration in the given
 // TcpProxy instance and builds a TCP filter out of it.
-func setAccessLogAndBuildTCPFilter(env *model.Environment, node *model.Proxy, config *tcp_proxy.TcpProxy) *listener.Filter {
-	setAccessLog(env, node, config)
+func setAccessLogAndBuildTCPFilter(meshConfig *meshconfig.MeshConfig, node *model.Proxy, config *tcp_proxy.TcpProxy) *listener.Filter {
```
Comment (costinm) [inline pilot/pkg/networking/core/v1alpha3/networkfilter.go]: @rshriram  I don't think this interface change is needed or good - passing env to all methods is better than passing just the mesh config ( even if methods happen to only use parts of env)

Comment (nmittler) [reply pilot/pkg/networking/core/v1alpha3/networkfilter.go]: So there were a couple of reasons that I chose to pass the mesh config around:

1. avoiding the (minor) overhead for the atomic read every time the mesh config is needed.
2. ensuring that we have a consistent view of the mesh config throughout the entire xds push. Before, it would be possible for the mesh config to change in the middle of a config push, which I suspect would be undesirable.

It's not clear to me what parts of the mesh config are allowed to be changed externally. It would be a lot simpler if the mesh config were just constant and we never had to worry about this.  In that case, just passing around Env would be a lot simpler.

Thoughts?

File: pilot/pkg/config/clusterregistry/multicluster.go (hunk 3 above)
Comment (nmittler) [reply pilot/pkg/config/clusterregistry/multicluster.go]: It's in the network watcher now.

File: pilot/pkg/bootstrap/mesh.go (hunk 1 above)
Comment (nmittler) [reply pilot/pkg/bootstrap/mesh.go]: The order is:

1. if a `*meshconfig.MeshConfig` was supplied in the arguments, use that.
2. if a file was provided and currently available, use that.
3. use a default mesh config.

File: pilot/pkg/config/coredatamodel/discovery.go (hunk 4 above)
Comment (nmittler) [reply pilot/pkg/config/coredatamodel/discovery.go]: I'm not entirely sure ... this PR just preserves the existing behavior.

File: pilot/pkg/bootstrap/mesh.go [hunk 6]
```diff
@@ -16,174 +16,87 @@ package bootstrap
...
+	if watcher == nil {
+		log.Info("mesh networks configuration not provided")
+		watcher = mesh.NewFixedNetworksWatcher(nil)
+	}
 
-// Add to the FileWatcher the provided file and execute the provided function
-// on any change event for this file.
-// Using a debouncing mechanism to avoid calling the callback multiple times
-// per event.
-func (s *Server) addFileWatcher(file string, callback func()) {
```
Comment (rshriram) [inline pilot/pkg/bootstrap/mesh.go]: where is this being moved to?

**Comment by rshriram** [comment]
I am not sure changing the mesh and networks to getters is a good idea, especially when the getters return the latest copy. If these configs change half way through a push to a single proxy, the proxy will potentially receive incorrect config. Or during a push epoch, it is important for the entire system to receive consistent config for debugging purposes. If some receive config based on updated mesh config and the rest receive old config, bad stuff happens. Also, we had code somewhere that would trigger a full push when the mesh or networks changed. That code is absolutely vital. I cant figure out from this PR if that old behavior is being preserved. Even if it is, it is all the more important to use the exact same mesh config/networks for all proxies during an epoch.

**Comment by nmittler** [comment]
@rshriram 

> If these configs change half way through a push to a single proxy, the proxy will potentially receive incorrect config...

But that's exactly how it works today. The watchers today just asynchronously overwrite the `Mesh` property of the environment, which is even worse since it could potentially lead to torn reads.

Perhaps we should just get rid of the file watchers altogether to avoid this problem?

@howardjohn @costinm

File: pilot/pkg/bootstrap/mesh.go (hunk 6 above)
Comment (nmittler) [reply pilot/pkg/bootstrap/mesh.go]: https://github.com/istio/istio/pull/19167/files#diff-94cf4f94e1d1aac6fe3fd24fbb526bfbR170

**Comment by nmittler** [comment]
@rshriram 

I've created https://github.com/istio/istio/pull/19340 to address your concern WRT config changing mid-push. PTAL

**Review by howardjohn — APPROVED** [review]
